Bugzilla/User.pm

```diff
@@ -2877,6 +2877,36 @@ sub user_id_to_login {
   return $login || '';
 }
 
+sub validate_password {
+  my $check = validate_password_check(@_);
+  ThrowUserError($check) if $check;
+  return 1;
+}
+
+sub validate_password_check {
+  my ($password, $matchpassword) = @_;
+
+  if (length($password) < USER_PASSWORD_MIN_LENGTH) {
+    return 'password_too_short';
+  }
+  elsif ((defined $matchpassword) && ($password ne $matchpassword)) {
+    return 'password_mismatch';
+  }
+
+  my $complexity_level = Bugzilla->params->{password_complexity};
+  if ($complexity_level eq 'bmo') {
+    my $features = 0;
+
+    $features++ if $password =~ /[a-z]/;
+    $features++ if $password =~ /[A-Z]/;
+    $features++ if $password =~ /[0-9]/;
+    $features++ if $password =~ /[^A-Za-z0-9]/;
+    $features++ if length($password) > 12;
```

Comment:
This is different from the stated requirements of "Password must contain at least one letter, a number and a special character".

A password of `abcdefghijkl1` passes the regex tests but shouldn't.

This also drops support for passphrases; we should continue to allow them.